File: test/four/app/src/main/java/com/z3d/scene/Scene3D.java
package com.z3d.scene;

import android.opengl.GLES20;

import com.z3d.base.Camera3D;
import com.z3d.base.ObjData;
import com.z3d.base.ObjDataManager;
import com.z3d.base.TexTuresBackFun;
import com.z3d.core.Context3D;
import com.z3d.display.Display3D;
import com.z3d.display.Display3DSprite;
import com.z3d.display.line.LineDisplayShader;
import com.z3d.display.role.Display3dMovie;
import com.z3d.filemodel.GroupDataManager;
import com.z3d.filemodel.MeshDataManager;
import com.z3d.filemodel.ParticleManager;
import com.z3d.filemodel.ResManager;
import com.z3d.filemodel.TextureManager;
import com.z3d.material.MaterialManager;
import com.z3d.material.TextureRes;
import com.z3d.program.ProgrmaManager;
import com.z3d.res.RoleRes;
import com.z3d.skill.SkillManager;
import com.z3d.units.AnimManager;
import com.z3d.units.LoadManager;
import com.z3d.units.TimeUtil;
import com.z3d.vo.Vector2D;
import com.z3d.vo.Vector3D;

import java.util.ArrayList;
import java.util.List;

public class Scene3D  {
    public ProgrmaManager progrmaManager;
    public Context3D context3D;
    public Camera3D camera3D;
    public Vector2D fogData;
    public Vector3D fogColor;
    public List<Display3D> displayList;
    public List<Display3dMovie> displayRoleList;
    public ParticleManager particleManager;
    public MeshDataManager meshDataManager;
    public TextureManager textureManager;
    public MaterialManager materialManager;
    public ObjDataManager objDataManager;
    public GroupDataManager groupDataManager;
    public LoadManager loadManager;
    public SkillManager skillManager;
    public AnimManager animManager;
    public ResManager resManager;
    public TextureRes pubLut;
    private float time;

    public Scene3D( ){
        this.time= TimeUtil.getTimer();
        this.context3D=new Context3D();
        this.camera3D=new Camera3D();
        this.displayList=new ArrayList<>();
        this.displayRoleList=new ArrayList<>();

        this.camera3D.rotationX =-30;
        this.camera3D.rotationY=45;
        this.loadManager=new LoadManager(this);
        this.particleManager=new ParticleManager(this);
        this.progrmaManager=new ProgrmaManager(this);
        this.meshDataManager=new MeshDataManager(this);
        this.textureManager=new TextureManager(this);
        this.groupDataManager=new GroupDataManager(this);
        this.skillManager=new SkillManager(this);
        this.materialManager=new MaterialManager(this);
        this.objDataManager=new ObjDataManager(this);
        this.animManager=new AnimManager(this);
        this.resManager=new ResManager(this);


    }
    public void  initData(){

        this.textureManager.getTexture( "base/brdf_ltu.jpg", new TexTuresBackFun() {
            @Override
            public void Bfun(TextureRes value) {
               pubLut =value;
            }
        });

        this.textureManager.getTexture( "base/brdf_ltu.jpg", new TexTuresBackFun() {
            @Override
            public void Bfun(TextureRes value) {
                 pubLut =value;
            }
        });
    }
    public  void  clearAll()
    {
        this.displayRoleList=new ArrayList<>();
        this.displayList=new ArrayList<>();
        particleManager.clearAll();
    }

    private void  upToGpu(){
        //纹理和着色

        this.progrmaManager.upDataProgramWaitIng();
        this.textureManager.upDataGenTextUserItem();
        this.objDataManager.upDataObjDataToGpu();
        this.resManager.upDataRoleResWaitIng();

    }
    public  void  upFrame()
    {

        upToGpu();

        Context3D ctx=this.context3D;
        ctx.setBlendParticleFactors(0);

        this.camera3D.upFrame();
        updateFrameRole();
        ctx.setFrontFace(true);
        ctx.setDepthTest(true);
        ctx.setWriteDepth(true);
        for(int i=0;  i< displayList.size();i++){
             displayList.get(i).upData();
        }
        for(int i=0; i< displayRoleList.size();i++){
            displayRoleList.get(i).upData();
        }
       
        ctx.setFrontFace(true);
        ctx.setDepthTest(false);
        ctx.setWriteDepth(false);
        this.skillManager.upData();
        this.particleManager.upFrame();


    }
    private void  updateFrameRole(){
        float _tempTime= TimeUtil.getTimer();
        float delay =  _tempTime - this.time;
        this.time=_tempTime;
        for(int i=0; i< displayRoleList.size();i++){
            displayRoleList.get(i).updateFrame(delay);
        }

    }

    public  void resizeScene()
    {
        GLES20.glViewport(0,0,this.camera3D.fovw,this.camera3D.fovh);
    }
    public  void  addDisplay(Display3D dis)
    {
        displayList.add(dis);
    }
    public  void  addMovieDisplay(Display3dMovie sc)
    {
        displayRoleList.add(sc);
    }


    public void addSpriteDisplay(Display3DSprite display) {
        addDisplay(display);
    }
}

File: h5gl/src/Pan3d/scene/Scene3D.ts
module Pan3d {
    export class Scene3D {
    
        public static getArrByStr(str: String): Array<string> {
            var boneNameAry: Array<string> = str.split(/\s+/g);
            for (var i: number = boneNameAry.length - 1; i >= 0; i--) {
                if (String(boneNameAry[i]).length < 1) {
                    boneNameAry.splice(i, 1);
                }
            }
            return boneNameAry;
        }
        public context3D: Context3D;
        public camera3D: Camera3D;
        public   skyCubeMap: Array<WebGLTexture>;
        public   pubLut: WebGLTexture;
        public get cam3D(): Camera3D {
            return this.camera3D
        }
        public get viewMatrx3D(): Matrix3D {
            return this.cam3D.cameraMatrix;
        }
        public   initPbr(): void {
            if (!this.pubLut) {
                this.textureManager.getTexture(this.fileRoot + "base/brdf_ltu.jpg", ($texture: TextureRes) => {
                    this.pubLut = $texture.texture;
                }, 1);
            }

            if (!this.skyCubeMap) {
                this.textureManager.loadCubeTexture(this.fileRoot + "base/cube/e", ($ary: any) => {
                    this.skyCubeMap = $ary;
                })
            }


        }
    
        public progrmaManager: ProgrmaManager;
        public objDataManager: ObjDataManager;
        public textureManager: TextureManager;
        public materialManager: MaterialManager;
        public particleManager: ParticleManager;
        public groupDataManager: GroupDataManager;
        public resManager: ResManager;
        public animManager: AnimManager;
        public meshDataManager: MeshDataManager;
        public skillManager: SkillManager;
        public fileRoot: string = "https://webpan.oss-cn-shanghai.aliyuncs.com/res/";
        public supportBlob: boolean;
        public fogColor: Array<number> = [0, 0, 0];
        public fogData: Array<number> = [1000, 0.003];
        public scaleLight: Array<number> = [2.0];
       
        public static frameTime: number = 1000 / 60;
        public static MAX_NUMBER: number = 10000000;
        protected _displayList: Array<Display3D>;
        protected _displayRoleList: Array<Display3dMovie>;
        constructor(value: WebGLRenderingContext) {

            this.supportBlob = true;
            this.context3D = new Context3D(value)
            this.camera3D = new Camera3D();
            this.progrmaManager = new ProgrmaManager(this);
            this.objDataManager = new ObjDataManager(this);
            this.textureManager = new TextureManager(this);
            this.materialManager = new MaterialManager(this);
            this.groupDataManager = new GroupDataManager(this);
            this.resManager = new ResManager(this);
            this.animManager = new AnimManager(this);
            this.meshDataManager = new MeshDataManager(this);
            this.particleManager = new ParticleManager(this);
            this.skillManager = new SkillManager(this);
            this._displayList = new Array();
            this._displayRoleList = new Array();
            this.initPbr();
            this.addDisplay(new GridLineSprite(this));
            // this.displayBaseSprite=new DisplayBaseSprite( this.context3D.webGlRender);
        }
        public clearAll() {
            this._displayList = new Array();
            this._displayRoleList = new Array();
            this.particleManager.clearAll();
        }
        private displayBaseSprite: DisplayBaseSprite;

        public addDisplay(itemDisplay: Display3D) {
            this._displayList.push(itemDisplay)
        }
        public addMovieDisplay(role: Display3dMovie) {
        this._displayRoleList.push(role);
        }

        public upFrame(): void {
            this.camera3D.upFrame();
            this.updateFrameRole();
            // this.camera3D.rotationY++;
            this.context3D.setBaseRender();
            this.context3D.setWriteDepth(true);
            this.context3D.setBlendParticleFactors(0);
            for (var i: number = 0; i < this._displayList.length; i++) {
                this._displayList[i].upFrame();
            }
            for (var i: number = 0; i < this._displayRoleList.length; i++) {
           
                this._displayRoleList[i].upFrame();
            }
            this.skillManager.update();
            this.particleManager.upFrame()
            this.displayBaseSprite ? this.displayBaseSprite.upFrame() : null;

        }
        public time:number=0;
        private   updateFrameRole():void{
            var _tempTime:number= TimeUtil.getTimer();
            var delay :number=  _tempTime - this.time;
            this.time=_tempTime;
            for (var i: number = 0; i < this._displayRoleList.length; i++) {
           
                this._displayRoleList[i].updateFrame(delay);
            }
            
    
        }
    

    }
}